### DISCUSSION TITLE
A few errors when running commands

### DISCUSSION THREAD

**Discussion by jgilbertfpv** [cmdwtf/Odiin#104] [Q&A] [answered]
I'm seeing `Board ID 1028 is not recognized [mbed_board]`, but if I run `pyocd flash -t nrf52840 mbr_nrf52_2.4.1_mbr.hex` from the `/components/softdevice/mbr/hex/` directory, It still seems to flash.
I was getting a fatal error referring to an unwanted or unused 'x' integer related to the amiitool version.c file and the `GIT_HASH_INT;` line but I believe I fixed by searching for mentions of `GIT_HASH_INT` in the makefile. On line 428 and 468, I removed the x and that seemed to solve the issue there. Next problem when trying to run make I encountered was this.
![Untitled](https://user-images.githubusercontent.com/37930433/227440997-9113fc35-dcdd-453c-85ba-4b1c52adfe1a.png)

**Comment by nitz** [comment]
Where are you seeing that "not recognized" board error? That sounds like it might be something that's changed in a tool that I need to take a look at.

For the fatfs error, I was working on updating it to a newer version that should get rid of that error. I just pushed a change with that, pull the latest and see if that works for you!

**Comment by jgilbertfpv** [comment]
Here's the other error I was seeing. What I did was remove the "x" from `CFLAGS += -DGIT_HASH_INT=0x$(GIT_COMMIT_HASH)` and `ASMFLAGS += -DGIT_HASH_INT=0x$(GIT_COMMIT_HASH)` in the makefile.
![Untitled](https://user-images.githubusercontent.com/37930433/228062554-c8ced46a-d404-4eac-9bdd-8bd1e1cbcc37.png)


Fatfs update fixed it! 
![Untitled1](https://user-images.githubusercontent.com/37930433/228063450-d15cc6ff-9c57-41e2-91e2-bcdb0acd205a.png)

When I try using make flash_mbr and make flash_all this is what I get back.
![Untitled](https://user-images.githubusercontent.com/37930433/228066890-9c7f865d-07eb-4738-b0fd-c64be06b132b.png)
Inside the bootloader/_build/build-odiin/ folder i do have a file called "odiin_bootloader-0.3.2-158-g23e6539.hex" and i notice the makefile is calling for one that ends with nosd.hex. could that be why the error above is showing?

**Comment by jgilbertfpv** [comment, marked as answer]
Ok so, I ran `make flash_mbr`, (after I ran make by itself) I moved both hex files to a new folder and ran `mergehex -m nrf52840_xxaa.hex odiin_bootloader-0.3.2-158-g23e6539.hex -o odiin.hex` then I ran `pyocd flash -t nrf52840 odiin.hex` and voilà!!
![20230327_181705](https://user-images.githubusercontent.com/37930433/228088467-66d1ee4f-50ad-4cdb-830b-5df4f97a7850.jpg)
The first time I didn't have to run `pyocd cmd -t nrf52840 --command write32 0xFF000 0x00000001 --command reset` but after doing some testing and trying to repeat the process, I did have to put that command after I flashed.

**Comment by nitz** [comment]
Heck yeah, love to hear it! I'll write a bit more about the other issues, but I'm glad you have workarounds that are going for you.

As for *actual* solutions to the other things you ran into:

- The `GIT_HASH_INT` bit, seems to have an potential issue on Windows if you don't have the gnu `head` on your path. If you've got Git for Windows installed in the normal location, try running `git rev-parse HEAD | "C:\Program Files\Git\usr\bin\head.exe" -c8`, and see if it just spits out 8 hex characters. I've so long had all of those sorts of GNU utils on my path that I forget somethings aren't part of Windows. I may need to add some duct tape to the makefile for it to work. Upside is, it's not used for anything important, just for version keeping.

- For 'board id 1028 not recognized' — It looks to me like `pyocd` may have done many changes, as I don't recall it having colored output before. Could you run `pyocd --version` and let me know what version it is? I probably just need to update and change which flags it uses.

- For the makefile wanting the nosd bootloader: It's _ideally_ supposed to use that one, since I'm not using any of the OTA flashing stuff. Using the version with the SD (SD is 'soft device', which is how Nordic distributes their bluetooth stack.) is just going to take up more flash space, which isn't an issue unless you hit the limit! (And with 1MB flash, the nrf52840 has plenty of room for activities.) As for why it's failing to create that one: the output you posted looks like `make` is failing to execute `python3` to run the `hexmerge.py` script that's in the bootloader repo. I'm betting `python3` is just not set up as an alias to `python`. I personally use `pyenv` to manage my python installs, and it creates that alias automatically so I hadn't run into it, but this is something that should be an easy makefile fix. In the meantime if you run `doskey python3=python $*` once before your run any of the make targets, it may just bop along happy!

**Reply by jgilbertfpv** [reply]
1. I do get `c69cdba3` back when i run the `git rev-parse HEAD | "C:\Program Files\Git\usr\bin\head.exe" -c8` I'll check into it a bit more.

2. Version `0.34.4.dev6` is what I get when I run `pyocd --verison`

3. I ran `doskey python3=python $*` and it didn't return anything,  I'll look into `pyenv` and see if I can get that error to go away. As for flash space, I'm not too concerned since I'm not getting any errors and tested out a few amiibos last night from the sd card and all seems to work just fine! Neat little device for sure!

On a side note- if it would be possible to add anything from this that has the option to randomize UID, and make multiple outputs of the same source-file for games like BOTW that will only allow you to scan in amiibos once per day, that could be a nice improvement. SD card would have plenty of space to make copies if there would be some way to run this from the device. Not sure if this project of yours is on the back burner at the moment or if you're still planning to make any future updates but might be worth a look. https://github.com/Lanjelin/AmiiboConverter
